File: 1. JPA basic/basic/src/main/java/study/jpa/basic/jpashop/domain/OrderItem.java
package study.jpa.basic.jpashop.domain;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

/**
 * @author Joonhyuck Hyoung
 */
@Entity(name = "jpashop_order_item")
@Table(name = "jpashop_order_item")
public class OrderItem {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long no;

    @ManyToOne
    @JoinColumn(name = "no")
    private Order order;

    @ManyToOne
    @JoinColumn(name = "no")
    private Item item;

    @Column(name = "order_price")
    private int orderPrice;
    private int count;

    public Long getNo() {
        return no;
    }

    public Order getOrder() {
        return order;
    }

    public void setOrder(Order order) {
        this.order = order;
    }

    public Item getItem() {
        return item;
    }

    public void setItem(Item item) {
        this.item = item;
    }

    public int getOrderPrice() {
        return orderPrice;
    }

    public int getCount() {
        return count;
    }
}

File: 2. JPA deep/jpashop/src/main/kotlin/jpabook/jpashop/domain/OrderItem.kt
package jpabook.jpashop.domain

import jpabook.jpashop.domain.item.*
import net.minidev.json.annotate.JsonIgnore
import javax.persistence.*

@Entity
data class OrderItem(
    @Id @GeneratedValue
    @Column(name = "order_item_id")
    val id: Long = 0L,

    /** 주문 상품 */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    val item: Item,

    /** 주문 */
    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    var order: Order? = null,

    /** 주문 가격 */
    val orderPrice: Int,

    /** 주문 수량 */
    val count: Int,
) {
    //== 비즈니스 로직 ==//
    /** 주문 취소 **/
    fun cancel() {
        item.addStock(count)
    }

    //== 조회 로직 ==//
    /** 주문상품 전체 가격 조회 **/
    fun getTotalPrice() = orderPrice + count

    //== 생성 메서드 ==//
    companion object {
        fun createOrderItem(item: Item, orderPrice: Int, count: Int): OrderItem {
            return OrderItem(
                item = item,
                orderPrice = orderPrice,
                count = count,
            ).also {
                it.item.removeStock(count)
            }
        }
    }
}
